src/shogun/machine/GLM.cpp

```diff
@@ -0,0 +1,170 @@
+	    ParameterProperties::HYPER, SG_OPTIONS(POISSON));
+	SG_ADD(
+	    &m_eta, "eta",
+	    "threshold parameter that linearizes the exp() function above eta",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_lambda, "lambda", "regularization parameter of penalty term",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_alpha, "alpha",
+	    "weighting between L1 penalty and L2 penalty term of the loss function",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_tolerance, "tolerance", "convergence threshold or stopping criteria",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_learning_rate, "learning_rate", "learning rate for gradient descent",
+	    ParameterProperties::HYPER);
+
+	m_gradient_updater = std::make_shared<GradientDescendUpdater>();
+	m_penalty = std::make_shared<ElasticNetPenalty>();
+	m_cost_function = std::make_shared<GLMCostFunction>();
+}
+
+GLM::GLM(
+    GLM_DISTRIBUTION distr, float64_t alpha, float64_t lambda,
+    float64_t learning_rate, int32_t max_iterations, float64_t tolerance,
+    float64_t eta)
+    : GLM()
+{
+	distribution = distr;
+	m_alpha = alpha;
+	m_lambda = lambda;
+	m_learning_rate = learning_rate;
+	m_max_iterations = max_iterations;
+	m_tolerance = tolerance;
+	m_eta = eta;
+
+	m_penalty->set_l1_ratio(m_alpha);
+}
+
+std::shared_ptr<RegressionLabels>
+GLM::apply_regression(std::shared_ptr<Features> data)
+{
+	if (data)
+	{
+		if (!data->has_property(FP_DOT))
+			error("Specified features are not of type CDotFeatures");
+		set_features(std::static_pointer_cast<DotFeatures>(data));
+	}
+
+	require(features, "Features are not provided");
+
+	auto num = features->get_num_vectors();
+	ASSERT(num > 0)
+	ASSERT(m_w.vlen == features->get_dim_feature_space())
+	SGVector<float64_t> out(num);
+	features->dense_dot_range(
+	    out.vector, 0, num, NULL, m_w.vector, m_w.vlen, bias);
+	auto result = m_cost_function->non_linearity(
+	    out, m_compute_bias, m_eta, distribution);
+
+	return std::make_shared<RegressionLabels>(result);
+}
+
+void GLM::init_model(const std::shared_ptr<Features> data)
+{
+	ASSERT(m_labels)
+	if (data)
+	{
+		if (!data->has_property(FP_DOT))
+			error("Specified features are not of type CDotFeatures");
+		set_features(std::static_pointer_cast<DotFeatures>(data));
+	}
+	ASSERT(features)
+
+	NormalDistribution<float64_t> normal_dist;
+	const auto& n_features = features->get_dim_feature_space();
+
+	if (m_w.vlen == 0)
+	{
+		if (m_compute_bias && bias == 0)
+			bias = 1.0 / (n_features + 1) * normal_dist(m_prng);
+
+		if (n_features > 0)
+		{
+			m_w = SGVector<float64_t>(n_features);
+
+			std::generate(m_w.begin(), m_w.end(), [&]() {
+				auto rand = normal_dist(m_prng);
+				return 1.0 / (n_features + 1) * rand;
+			});
+		}
+	}
+}
+
+void GLM::iteration()
+{
+	SGVector<float64_t> w_old = m_w.clone();
+
+	auto X = get_features()->get_computed_dot_feature_matrix();
+	auto y = regression_labels(get_labels())->get_labels();
```

Comment:
this should be fixed at some point. I guess when we have the LabelEncoder in Machine, otherwise you are performing a potentially expensive operation in each iteration

Reply:
it seems like that `RegressionLabels` is what LabelEncoder lacks, currently, LabelEncoder only support MulticlassLabels and BinaryLabels.

Reply:
True, but it should be possible to add it?